3.2 backports 202310

vadz opened the pull request.

vadz committed:
Fix events generated when toast-based notifications are dismissed

Make wxToastEventHandler::QueryInterface() work correctly by
implementing it ourselves instead of relying on the stuff in
wx/msw/ole/comimpl.h which only works for the classes supporting a
single (not counting IUnknown) interface, as the object pointer has to
be cast to the pointer to the correct interface when there is more than
one of them.

See #24320, #24826.

(cherry picked from commit 38132bcb945d2b9f33f3fcd54328fd8c6e013eb9)

lanurmi committed:
Remove unused wxString variables

A few were removed tests/, which might have existed for readability.

(cherry picked from commit 007787c3a8608ce2e16122445c478b55f64f99c9)

vadz committed:
Fix position of the icons when printing in wxMSW

We need to account for wxPrinterDC origin when drawing the icons, so use
the same XLOG2DEV() and YLOG2DEV() macros as in src/msw/dc.cpp in this
file and apply them to the coordinates before passing them to the native
functions.

See #24673, #24834.

(cherry picked from commit 6ecc7af12fe5be29bb1aeb3fddfe2c57efb8550b)

lanurmi committed:
Add macro wxWARN_UNUSED to stand for the attribute 'warn_unused'

Also known as [[gnu::warn_unused]].

Use __has_attribute to check availability of the attribute. (Is available
on recent GCC and Clang versions, not available on MSVC.)

(cherry picked from commit a2e2bc57c9fd064503884b91af25ad4335263379)

lanurmi committed:
Enable the possibility to warn about unused wxStrings

Use the wxWARN_UNUSED attribute in wxString's definition. Compilers do not
typically warn about unused non-POD objects, but this attribute makes it
possible (on GCC and Clang).

Whether a warning is actually shown or not still depends on whether
the user has enabled warnings for unused variables in their compiler flags.

(cherry picked from commit a73504cdc5433a8a89a3dd2e9decc325735bba6b)

vadz committed:
Move code logging GDK atom being added to clipboard

No real changes, just put this log into the function actually doing this
instead of doing it in one place where it's called.

(cherry picked from commit 20a17c0ef9b935abbb43a777d39bb5dba0d1663e)

vadz committed:
Extract wxAtomName() helper function

No real changes, just simplify calls to gdk_atom_name().

(cherry picked from commit d7ec3a1502363cc01ec09ff79a8d41249d281a1c)

vadz committed:
Use gtk_selection_add_targets() in wxGTK clipboard code

For some reason, calling gtk_selection_add_target() multiple times
doesn't work under some Wayland compositors, including at least Weston:
only the first call to it is taken into account, but the other atoms
don't appear to be added to the list of supported formats.

Work around this by calling gtk_selection_add_targets() to add all of
them at once, which works with at least Weston.

See #24701.

(cherry picked from commit 0373d60a36b8b7a03e163a7134ccafb1642ad5b8)

vadz committed:
Define helper "isWayland" variable in wxClipboard::AddData()

No real changes, just make the code a bit more readable.

(cherry picked from commit 56ed13dfb76a7e8dc64d8c1d277619d4eb36a0b5)

vadz committed:
Don't bother advertising TIMESTAMP under Wayland

This doesn't seem to be used there, and no native Wayland applications
ever put it on clipboard, so don't do it neither when using Wayland.

(cherry picked from commit 4c83451b76480f5ac113ee2d88c89ceca56e1532)

vadz committed:
Always provide Wayland native clipboard text formats ourselves

Don't rely on GTK for doing it as this has been only added in 3.24.35
and there are systems using Wayland with older GTK versions, e.g. Ubuntu
22.04 which uses 3.24.33.

Also put Wayland native formats first, before the X11 ones: even though
it doesn't seem to really matter, it's again more consistent with what
the native Wayland applications do.

See #24701.

(cherry picked from commit a24b3dd233099884b66d7f3a8a4c8b6f46c05a24)

vadz committed:
Revert hack with ignoring wxDF_TEXT under Wayland

This reverts 1b3afc41db (Don't advertise wxDF_TEXT when copying text
under Wayland, 2024-03-20) which isn't necessary any longer after the
changes of eeda1f14f5 (Use gtk_selection_add_targets() in wxGTK
clipboard code, 2024-09-28) as adding all targets at once avoids the
problem with STRING replacing UTF8_STRING etc.

(cherry picked from commit c2a2d270417552c130c965fb0fde6a50a8d50f52)

vadz committed:
Order clipboard formats in Wayland-first order under Wayland

This doesn't seem to make any real difference, but is again more
consistent with the way the native Wayland applications work.

(cherry picked from commit ecdc3b2fbc327a20c21ef297a3275d327590f0c2)

vadz committed:
Document Wayland clipboard fixes in the change log

vadz committed:
Don't give error in docs workflow for PRs

Just skip the upload step for them.

MaartenBent committed:
Cleanup and reorder wrapcdlg.h includes

Include wrappers in a specific order so missing.h does not cause macro redefinition warnings.
When wrapcdlg.h/<commdlg.h> is not needed, include wx/msw/private.h instead.

See #24851, #24860.

(cherry picked from commit b170a58b403a51a25a973d84abe4b490fa5b3cac)

lanurmi committed:
Use unused variables like they were likely meant to be used

No actual changes to the test, a slight change of tone with one colour
scheme in the propgrid sample.

(cherry picked from commit ee978a1f6f873d67571e2cc5634d1785389649cf)

vadz committed:
Fixes for autoconf build with gcc 14 on Solaris

See #24869.

(cherry picked from commit 1880fe8b64bb34713f9baf27abd038b50d8dc558)

lanurmi committed:
Add the wxWARN_UNUSED attribute for wxStringTokenizer; remove unused

(cherry picked from commit a76d6642835a10d16741ef13c906064a512a01d5)

lanurmi committed:
Add the wxWARN_UNUSED attribute for wxArrayString

(cherry picked from commit 55f0b7f9a858079f5856e5545035348e20037029)

lanurmi committed:
Add the wxWARN_UNUSED attribute for wxSize, wxRect, wx(Real)Point; remove unused

(cherry picked from commit 1d0e1b69cd0d366217879df8f9614aa7716b6f8d)

lanurmi committed:
Add the wxWARN_UNUSED attribute for wxLongLong, wxULongLong

(cherry picked from commit 0a75b8a4b792ef0ceff17ee32b5ab4c9c7f2b936)

lanurmi committed:
Add the wxWARN_UNUSED attribute for wxColour; remove unused

(cherry picked from commit 463d903b2339f8c8da55955e5e3259fc2027d456)

lanurmi committed:
Add the wxWARN_UNUSED attribute for wxDateTime

(cherry picked from commit 00df4754fdc80eee12e43944a3481a8329b13351)

vadz committed:
Fix always using "All pages" in wxGTK wxPrintDialog

The code in wxGtkPrinter tried to determine whether all pages should be
printed from the values returned by GetPageInfo(), but this doesn't make
much sense because a typical implementation of GetPageInfo() uses the
values in the associated wxPrintDialogData to determine the pages to
return, i.e. it's wxPrintDialogData which is the primary source of truth
and the values set in it should be respected, otherwise we ended up with
"All pages" being selected in the print dialog independently of the
values set in the application code.

Just remove this code, which isn't present in the other ports neither,
to fix this.

See #24868.

(cherry picked from commit 7abc628a5ab6084836532a5e2f018758c4cf386d)

vadz committed:
Handle passing root item to wxTreeListCtrl::Select() gracefully

Assert if an invalid item is passed to Select(), Unselect() or
IsSelected() instead of crashing.

See #24861.

(cherry picked from commit 9b1231e4ba6742e1200c38edb1ba0d3f9c8ee438)

lanurmi committed:
Make wxWARN_UNUSED an opt-in feature in 3.2

Define wxWARN_UNUSED always as nothing, unless wxNO_UNUSED_VARIABLES is
defined by the library user.

Also opt-in for wxWARN_UNUSED when building wxWidgets itself, i.e. when
WXBUILDING is defined.

csomor committed:
Fix wxOSX wxScreenDC compilation against SDK 15.0

Don't use CGDisplayCreateImage() which is not available in macOS 15 any
more.

See #24724, #24879.

(cherry picked from commit 7e32de0e7eb8df37f3cf3c78d4f736fc7111f899)

PBfordev commented:
On wxForum, I encountered SVGs that are rasterized incorrectly in 3.2 but correctly in the master, probably due to difference in bundled nanosvg library.

Is there a reason why nanosvg is not being updated in 3.2 branch?

swt2c committed:
Update list of locales to install

The list of the languages to install the message catalogs for has to be
hardcoded in the bakefile currently, so update it to ensure that all the
available catalogs are installed.

See #24880, #24887.

(cherry picked from commit fb04f7c2cc3e07a6de4e94747fe38099b60678bc)

vadz committed:
Merge branch 'warn-unused-3.2' into 3.2

Backport wxWARN_UNUSED into 3.2, along with removing unused variables.

See #24882.

vadz commented:
> Is there a reason why nanosvg is not being updated in 3.2 branch?

I don't think there is some special reason, it's just that somebody needs to do it, i.e. cherry pick 19aee81c1d (Update nanosvg to the latest upstream version, 2024-01-14) into 3.2 (should be simple, unless I'm forgetting something) and check that it works there.

vadz committed:
Fix height of wxOwnerDrawnComboBox popup at different DPI

See #24850, #24898.

(cherry picked from commit 0150248f08ed4687fbf8889e058261f0c741fd14)

lanurmi committed:
Make wxWARN_UNUSED work with Clang (again)

Checking for availability of 'warn_unused' using __has_cpp_attribute
evaluates to false with Clang, even when the attribute is actually
available. __has_attribute gives the correct result, so use it instead,
despite the aesthetic inconsistency with other checks.

See #24833, #24907.

(cherry picked from commit 795c02c4982c1d8f3e192b3e3e3ffa70193bb72e)

lanurmi committed:
Fix compilation of <wx/archive.h> on Clang 19 on Fedora

Clang 19.1.0 on Fedora Rawhide fails to compile <wx/archive.h>
because m_rep's members are accessed as if m_rep was not a pointer,
but it is.

This is obviously wrong with any compiler, but apparently because
wxArchiveIterator is a template class, and its operator=() is
never called in tests, its brokenness went unnoticed.

(cherry picked from commit c70a281c335d3df773b3ce4b5149aed92c9ee8a7)

See #24922.

lanurmi commented:
Requesting 795c02c.
